### PR TITLE
[Merged by Bors] - feat: port MeasureTheory.Function.LpSpace

### PR DESCRIPTION
**Pull request by Komyyy**
---

- [x] depends on: https://github.com/leanprover-community/mathlib/pull/19083
- [x] depends on: #4379
- [x] depends on: #4445

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by Komyyy** [comment]
I will sleep.

**Comment by eric-wieser** [comment]
Please don't port this file, I have a substantial refactor in the works that is on the porting dashboard.

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib#19083
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by eric-wieser** [comment]
@Komyyy,rather than spendnig more time on this PR, it might be worth thinking about how to split the file in mathlib3 (waiting for https://github.com/leanprover-community/mathlib/pull/19083 first)

**Comment by Komyyy** [comment]
@eric-wieser I don't know about ecosystems in Mathlib 3 as much as Mathlib 4, so I want to specialize Mathlib 4.
Since many files depends on this file, I think it's important to work for this file in advance.

**Comment by eric-wieser** [comment]
> I think it's important to work for this file in advance.

If the porting dashboard says "please wait for mathlib3#NNNN" this means "do not work on this file in advance (at least, not without starting a Zulip thread) because your work may be wasted when the file changes"

**Comment by Komyyy** [comment]
@eric-wieser I see.

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib#19083~~
* ~~leanprover-community/mathlib4#4379~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
Eric's suggestion is that we reopen this when mathport next produces fresh output. Sorry, @Komyyy, for your wasted work here. We are trying to avoid having porting targets wait on mathlib3 PRs as much as possible, but unfortunately we are not quite there yet.

**Comment by eric-wieser** [comment]
Once #4445 is in, please rebase this on the latest mathport output so that I can confirm it is up to date with mathlib3 (or ping me and I will do it)

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib#19083~~
* ~~leanprover-community/mathlib4#4379~~
* ~~leanprover-community/mathlib4#4445~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Review by eric-wieser — COMMENTED** [review]
The commit history has gotten rather long here, but the latest diff is

https://github.com/leanprover-community/mathlib4/pull/4341/files/9d7132edc4e4570fc27abf7c30cf8959cad1bc1b..5a49fa44d713ccfed517b096894e41e5168714ca

File: Mathlib/MeasureTheory/Function/LpSpace.lean [hunk 1]
```diff
@@ -13,7 +13,7 @@ import Mathlib.MeasureTheory.Function.LpSeminorm
 import Mathlib.Topology.ContinuousFunction.Compact
 
 /-!
-# Lp space
+# ℒp space and Lp space
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Function/LpSpace.lean:16]: This isn't synchronized correctly

File: Mathlib/MeasureTheory/Function/LpSpace.lean [hunk 2]
```diff
@@ -593,42 +611,43 @@ theorem snorm_indicator_const {c : G} (hs : MeasurableSet s) (hp : p ≠ 0) (hp_
...
-  simp_rw [snorm_ess_sup, nnnorm_indicator_eq_indicator_nnnorm, ENNReal.coe_indicator]
+  simp_rw [snormEssSup, nnnorm_indicator_eq_indicator_nnnorm, ENNReal.coe_indicator]
   by_cases hs_null : μ s = 0
-  · rw [measure.restrict_zero_set hs_null]
+  · rw [Measure.restrict_zero_set hs_null]
     simp only [essSup_measure_zero, ENNReal.essSup_eq_zero_iff, ENNReal.bot_eq_zero]
-    have hs_empty : s =ᵐ[μ] (∅ : Set α) := by rw [ae_eq_set]; simpa using hs_null
+    have hs_empty : s =ᵐ[μ] (∅ : Set α) := by
+      rw [ae_eq_set]
+      simpa using hs_null
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Function/LpSpace.lean]: May as well put this back on one line

File: Mathlib/MeasureTheory/Function/LpSpace.lean [hunk 3]
```diff
@@ -639,58 +658,61 @@ theorem snorm_indicator_eq_snorm_restrict {f : α → F} (hs : MeasurableSet s)
...
+  suffices (∫⁻ x, (‖s.indicator f x‖₊ : ℝ≥0∞) ^ p.toReal ∂μ) =
+      ∫⁻ x in s, (‖f x‖₊ : ℝ≥0∞) ^ p.toReal ∂μ by rw [this]
   rw [← lintegral_indicator _ hs]
   congr
   simp_rw [nnnorm_indicator_eq_indicator_nnnorm, ENNReal.coe_indicator]
-  have h_zero : (fun x => x ^ p.to_real) (0 : ℝ≥0∞) = 0 := by
+  have h_zero : (fun x => x ^ p.toReal) (0 : ℝ≥0∞) = 0 := by
     simp [ENNReal.toReal_pos hp_zero hp_top]
-  exact (Set.indicator_comp_of_zero h_zero).symm
+  exact (Set.indicator_comp_of_zero (g := fun x : ℝ≥0∞ => x ^ p.toReal) h_zero).symm
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Function/LpSpace.lean:670]: This one is worth a porting note I think, assuming it's still needed.

File: Mathlib/MeasureTheory/Function/LpSpace.lean [hunk 4]
```diff
@@ -718,97 +740,100 @@ theorem indicatorConstLp_coeFn : ⇑(indicatorConstLp p hs hμs c) =ᵐ[μ] s.in
...
     (hf : StronglyMeasurable f) (hg : StronglyMeasurable g) :
     Memℒp (f + g) p μ ↔ Memℒp f p μ ∧ Memℒp g p μ := by
   borelize E
   refine' ⟨fun hfg => ⟨_, _⟩, fun h => h.1.add h.2⟩
-  · rw [← indicator_add_eq_left h]; exact hfg.indicator (measurableSet_support hf.measurable)
-  · rw [← indicator_add_eq_right h]; exact hfg.indicator (measurableSet_support hg.measurable)
+  · rw [← Set.indicator_add_eq_left h]
+    exact hfg.indicator (measurableSet_support hf.measurable)
+  · rw [← Set.indicator_add_eq_right h]
+    exact hfg.indicator (measurableSet_support hg.measurable)
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Function/LpSpace.lean]: Again, I don't think these lines need to be split.

File: Mathlib/MeasureTheory/Function/LpSpace.lean [hunk 5]
```diff
@@ -922,118 +947,121 @@ namespace ContinuousLinearMap
...
+theorem coeFn_compLpL [Fact (1 ≤ p)] (L : E →L[𝕜] F) (f : Lp E p μ) :
     L.compLpL p μ f =ᵐ[μ] fun a => L (f a) :=
   L.coeFn_compLp f
 #align continuous_linear_map.coe_fn_comp_LpL ContinuousLinearMap.coeFn_compLpL
 
 theorem add_compLpL [Fact (1 ≤ p)] (L L' : E →L[𝕜] F) :
-    (L + L').compLpL p μ = L.compLpL p μ + L'.compLpL p μ := by ext1 f; exact add_comp_Lp L L' f
+    (L + L').compLpL p μ = L.compLpL p μ + L'.compLpL p μ := by
+  ext1 f
+  exact add_compLp L L' f
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Function/LpSpace.lean]: Leave on one line?

File: Mathlib/MeasureTheory/Function/LpSpace.lean [hunk 6]
```diff
@@ -922,118 +947,121 @@ namespace ContinuousLinearMap
...
 #align continuous_linear_map.add_comp_LpL ContinuousLinearMap.add_compLpL
 
+set_option synthInstance.maxHeartbeats 30000 in
 theorem smul_compLpL [Fact (1 ≤ p)] {𝕜'} [NormedRing 𝕜'] [Module 𝕜' F] [BoundedSMul 𝕜' F]
     [SMulCommClass 𝕜 𝕜' F] (c : 𝕜') (L : E →L[𝕜] F) : (c • L).compLpL p μ = c • L.compLpL p μ := by
-  ext1 f; exact smul_comp_Lp c L f
+  ext1 f
+  exact smul_compLp c L f
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Function/LpSpace.lean]: idem

File: Mathlib/MeasureTheory/Function/LpSpace.lean [hunk 7]
```diff
@@ -1204,77 +1230,78 @@ theorem snorm_lim_le_liminf_snorm {E} [NormedAddCommGroup E] {f : ℕ → α →
...
   exact snorm_ne_top _
-#align measure_theory.Lp.cauchy_seq_Lp_iff_cauchy_seq_ℒp MeasureTheory.lp.cauchySeq_lp_iff_cauchySeq_ℒp
+#align measure_theory.Lp.cauchy_seq_Lp_iff_cauchy_seq_ℒp MeasureTheory.Lp.cauchySeq_Lp_iff_cauchySeq_ℒp
 
 theorem completeSpace_lp_of_cauchy_complete_ℒp [hp : Fact (1 ≤ p)]
     (H :
       ∀ (f : ℕ → α → E) (hf : ∀ n, Memℒp (f n) p μ) (B : ℕ → ℝ≥0∞) (hB : (∑' i, B i) < ∞)
         (h_cau : ∀ N n m : ℕ, N ≤ n → N ≤ m → snorm (f n - f m) p μ < B N),
-        ∃ (f_lim : α → E)(hf_lim_meas : Memℒp f_lim p μ),
+        ∃ (f_lim : α → E), Memℒp f_lim p μ ∧
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Function/LpSpace.lean]: Is there a reason that you changed the statement here?

File: Mathlib/MeasureTheory/Function/LpSpace.lean [hunk 8]
```diff
@@ -1541,7 +1570,7 @@ end CompleteSpace
 /-! ### Continuous functions in `Lp` -/
 
 
-open scoped BoundedContinuousFunction
+open BoundedContinuousFunction
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Function/LpSpace.lean]: This looks like an error to me

File: Mathlib/MeasureTheory/Function/LpSpace.lean [hunk 9]
```diff
@@ -1553,112 +1582,115 @@ variable (E p μ)
...
+    ((toLpHom p μ).range : AddSubgroup (Lp E p μ)) =
+      MeasureTheory.Lp.boundedContinuousFunction E p μ := by
   symm
   convert AddMonoidHom.addSubgroupOf_range_eq_of_le
-      ((ContinuousMap.toAEEqFunAddHom μ).comp (to_continuous_map_add_hom α E))
-      (by rintro - ⟨f, rfl⟩; exact mem_Lp f : _ ≤ Lp E p μ)
+      ((ContinuousMap.toAEEqFunAddHom μ).comp (toContinuousMapAddHom α E))
+      (by
+        rintro - ⟨f, rfl⟩
+        exact mem_Lp f : _ ≤ Lp E p μ)
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Function/LpSpace.lean]: un-wrap?

**Review by eric-wieser — APPROVED** [review]
The commit history has gotten rather long here, but the latest diff is

https://github.com/leanprover-community/mathlib4/pull/4341/files/9d7132edc4e4570fc27abf7c30cf8959cad1bc1b..5a49fa44d713ccfed517b096894e41e5168714ca

This looks pretty good; my main comment is that porting a stack of 3 or 4 PRs like this all at once means:
* The later PRs are hard to review because there are 50 commits to navigate around
* You end up using a version of the file from an older `mathport` binary, meaning you don't get the fixes for `open` vs `open scoped`, or any name fixes introduced by missing `#align`s elsewhere in mathlib4.

In future, if you make these sorts of PR chain,  once a PR becomes unblocked please re-compare against the latest mathport output (by running `start_port.sh` again and rebasing/merging) to make sure you've picked up any fixes that were made in the meantime.

File: Mathlib/MeasureTheory/Function/LpSpace.lean (hunk 7 above)
Comment (Komyyy) [reply Mathlib/MeasureTheory/Function/LpSpace.lean]: In Lean 4, the bounded existential quantifer `∃ x ∈ s, p x` is elaborated to `∃ x, x ∈ s ∧ p x` while it's elaborated to `∃ x (_ : x ∈ s), p x`. Following the way, some props like `∃ x (_ : p x), q x` is ported to `∃ x, p x ∧ q x` like `[isPiSystem_Ioo](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/PiSystem.html#isPiSystem_Ioo)`.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**